### PR TITLE
Add intrinsic for KCallable.name property

### PR DESCRIPTION
**Pull request by cypressious**
Fixes #KT-11531

### CODE REVIEW

File: compiler/testData/codegen/box/reflection/properties/propertyName.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    val a = ""
+    val b = ""
+}
+
+fun box(): String {
+    val a = A::a.name
+    if (a != "a") return "Fail $a"
+    val b = A::`b`.name
+    if (b != "b") return "Fail $b"
+    return "OK"
```
Comment (udalov) [inline compiler/testData/codegen/box/reflection/properties/propertyName.kt]: This test doesn't check that no classes were generated for the property reference. I don't think it's possible in a `box` test, please try a bytecode listing test instead (`compiler/testData/codegen/bytecodeListing`).

Also please test constructors and functions

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KCallableNameProperty.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.jetbrains.org.objectweb.asm.Type
+
+class KCallableNameProperty : IntrinsicPropertyGetter() {
+    override fun generate(resolvedCall: ResolvedCall<*>?, codegen: ExpressionCodegen, returnType: Type, receiver: StackValue): StackValue? {
+        val expressionReceiver = resolvedCall!!.dispatchReceiver as? ExpressionReceiver ?: return null
+        val expression = expressionReceiver.expression as? KtCallableReferenceExpression ?: return null
+        val callableReference = expression.callableReference
+        val descriptor = callableReference.getResolvedCall(codegen.bindingContext)?.resultingDescriptor ?: return null
+
+        val name = if (descriptor is ConstructorDescriptor) {
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KCallableNameProperty.kt]: `descriptor.name.asString()`

File: compiler/testData/codegen/box/reflection/properties/propertyName.kt (hunk 1 above)
Comment (cypressious) [inline compiler/testData/codegen/box/reflection/properties/propertyName.kt]: It seems classes are indeed generated. Where should I look to prevent that?

Comment (cypressious) [inline compiler/testData/codegen/box/reflection/properties/propertyName.kt]: Should I hook into `ExpressionCodegen#visitSimpleNameExpression` and prevent the anonymous class from being generated or is there an approach where I can remove the generated class afterwards?

**Comment by cypressious** [comment]
Constructor test already exists in `compiler\testData\codegen\box\reflection\constructors\constructorName.kt`

File: compiler/testData/codegen/box/reflection/properties/propertyName.kt (hunk 1 above)
Comment (cypressious) [inline compiler/testData/codegen/box/reflection/properties/propertyName.kt]: I got it working by modifying `visitSimpleNameExpression`.

**Comment by udalov** [comment]
> Constructor test already exists in `compiler\testData\codegen\box\reflection\constructors\constructorName.kt`

`KCallable#name` itself is already tested, and not only for constructors but for properties and functions as well. I think I was wrong in suggesting you to use `codegen/box` tests, they won't test anything here, so maybe they should be deleted.

**Comment by cypressious** [comment]
I'll remove the tests. Any feedback on the implementation?

**Comment by udalov** [comment]
I think the implementation is fine, could not think of anything substantially better. I've launched a remote run with your changes.

**Comment by udalov** [comment]
Merged (https://github.com/JetBrains/kotlin/commit/4a619db7219e1a09b63c27d28db25b44ae10066c), thank you for the contribution!

**Pull request closed without merging**